src/prefect/utilities/threaded_heartbeat.py

```diff
@@ -0,0 +1,65 @@
+import logging
+import threading
+from contextlib import contextmanager
+from typing import Iterator
+
+import prefect
+from prefect import config
+from prefect.client import Client
+from prefect.utilities.logging import get_logger
+
+
+class HeartbeatThread(threading.Thread):
+    def __init__(
+        self: "HeartbeatThread",
+        stop_event: "threading.Event",
+        flow_run_id: str,
+        num: int = None,
+    ) -> None:
+        threading.Thread.__init__(self)
+        # 'daemonizes' the thread, so it will terminate when all non-daemonized threads have finished
+        self.daemon = True
+        self.flow_run_id = flow_run_id
+        self.num = num
+        self.stop_event = stop_event
+
+    def run(self: "HeartbeatThread") -> None:
```

Comment:
Minor note that `mypy` doesn't care if you don't annotate `self` on a class method